### PR TITLE
add prerun and postrun shell hooks for keadm edge commands

### PR DESCRIPTION
**Pull request by tangming1996**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Previously, we discussed in the issue https://github.com/kubeedge/kubeedge/issues/5745 about automatically installing the container runtime before executing the keadm join command. The conclusion is that there are many implementations and versions of container runtimes, and it will be a huge effort for the KubeEdge community to maintain them.

Therefore, we thought of using prerun and postrun shell hooks to handle container runtime installation and uninstallation, as well as more flexible applications. E.g., keadm join --preun install-constainerd.sh, keadm reset --postrun uninstall-containerd.sh.
**Which issue(s) this PR fixes**:

Fixes a part of https://github.com/kubeedge/kubeedge/issues/5940

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by luomengY** [comment]
Let `e2e` pass the test first

**Comment by tangming1996** [comment]
> Let `e2e` pass the test first让 `e2e` 先通过测试

OK

File: keadm/cmd/keadm/app/cmd/util/scriptrun.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+package util
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+)
+
+func RunScript(scriptPath string) error {
```
Comment (WillardHu) [inline keadm/cmd/keadm/app/cmd/util/scriptrun.go:25]: This function is very simple now, I don't think it needs to be abstracted yet. In keadm, the `exec.Command` is used in many places, We can cansider how to abstract them later.

File: keadm/cmd/keadm/app/cmd/edge/join_others.go [hunk 2]
```diff
@@ -181,6 +184,12 @@ func createEdgeConfigFiles(opt *common.JoinOptions) error {
 }
 
 func join(opt *common.JoinOptions, step *common.Step) error {
+	if opt.PreRun != "" {
+		step.Printf("Execute the prescript before joining the node")
+		if err := util.RunScript(opt.PreRun); err != nil {
+			return err
+		}
+	}
```
Comment (WillardHu) [inline keadm/cmd/keadm/app/cmd/edge/join_others.go:192]: Move to PreRunE，and add `--post-run` flag. Reset need add `--pre-rune` flag too.
```goalng
PreRunE: func(cmd *cobra.Command, args []string) error {
    ...
    if opt.PreRun != "" {
        // run script
    }
}
PostRun: func(cmd *cobra.Command, args []string) error {
    ....
    if opt.PostRun != "" {
        // run script without err handle
    }
}
```

Comment (tangming1996) [reply keadm/cmd/keadm/app/cmd/edge/join_others.go]: Thank you for your suggestion. I have modified it.

File: keadm/cmd/keadm/app/cmd/common/constant.go [hunk 3]
```diff
@@ -50,6 +50,9 @@ const (
 
 	// FlagNamePrintFinalValues ...
 	FlagNamePrintFinalValues = "print-final-values"
+
+	// FlagNamePostRun ...
+	FlagNamePostRun = "post-run"
```
Comment (fujitatomoya) [inline keadm/cmd/keadm/app/cmd/common/constant.go:55]: why is this assigned to `Cloud init` flag constant? according to https://github.com/kubeedge/kubeedge/issues/5940, it is supposed to be under `Edge join flag names`? Or if those are common flags, they can be under `Common flag names`?

File: keadm/cmd/keadm/app/cmd/edge/join.go [hunk 4]
```diff
@@ -97,6 +103,15 @@ func NewEdgeJoin() *cobra.Command {
 
 			return nil
 		},
+		PostRunE: func(cmd *cobra.Command, args []string) error {
+			if joinOptions.PostRun != "" {
+				step.Printf("Execute the prescript before joining the node")
```
Comment (fujitatomoya) [inline keadm/cmd/keadm/app/cmd/edge/join.go:108]:
```suggestion
				step.Printf("Execute the postscript after joining the node")
```

File: keadm/cmd/keadm/app/cmd/edge/reset_others.go [hunk 5]
```diff
@@ -51,6 +51,12 @@ func NewOtherEdgeReset() *cobra.Command {
 		Long:    resetLongDescription,
 		Example: resetExample,
 		PreRunE: func(cmd *cobra.Command, args []string) error {
+			if reset.PreRun != "" {
+				fmt.Printf("Execute the prescript before joining the node")
```
Comment (fujitatomoya) [inline keadm/cmd/keadm/app/cmd/edge/reset_others.go:55]:
```suggestion
				fmt.Printf("Execute the prescript before resetting the node")
```

**Review by fujitatomoya — COMMENTED** [review]
IMO pre/post script hook is generally useful to do user specific things.

a few minor comments need to be addressed.

btw,

> https://github.com/kubeedge/kubeedge/issues/5745
> surport container runtime auto install

if that is the major case, we would want to provide the example script to install container runtime? this would be useful against the original issue description?

File: keadm/cmd/keadm/app/cmd/common/constant.go (hunk 3 above)
Comment (tangming1996) [reply keadm/cmd/keadm/app/cmd/common/constant.go]: Thank you for your suggestion. I have modified it.

File: keadm/cmd/keadm/app/cmd/edge/join.go (hunk 4 above)
Comment (tangming1996) [reply keadm/cmd/keadm/app/cmd/edge/join.go]: Thank you for your suggestion. I have modified it.

File: keadm/cmd/keadm/app/cmd/edge/reset_others.go (hunk 5 above)
Comment (tangming1996) [reply keadm/cmd/keadm/app/cmd/edge/reset_others.go]: Thank you for your suggestion. I have modified it.

**Comment by tangming1996** [comment]
> IMO pre/post script hook is generally useful to do user specific things.IMO 前/后脚本钩子通常对于执行用户特定的事情很有用。
> 
> a few minor comments need to be addressed.一些小的评论需要解决。
> 
> btw, 增值税
> 
> > #5745
> > surport container runtime auto installsurport 容器运行时自动安装
> 
> if that is the major case, we would want to provide the example script to install container runtime? this would be useful against the original issue description?如果这是主要情况，我们是否要提供示例脚本来安装 Container Runtime？这对原始问题描述有用吗？

Supporting the container runtime is just a scenario we assume. There may be other requirements from users, such as environment checks and so on. The specific scripts needed should be customized according to the user's own scenarios. Of course, you are quite right. It is very necessary to provide a case. Later, we will submit a separate pull request (pr) to provide a case of installing the container runtime.

**Pull request merged**